File: server_python/amq/handlers.py
from amq import sio, db

class ClientUpdater(object):

    @staticmethod
    def sendNotification(msgType, msg, sid=None):
        sio.emit('SYSTEM_NOTIFICATION', data=(msgType, msg), room=sid)

    @staticmethod
    def updateUserList(user, sid=None):
        data = {
            'list': db.getUserList(user),
            'file': user
        }
        sio.emit('SYNC_USER_LIST', data, room=sid)

    @staticmethod
    def updateUsers(sid=None):
        sio.emit('SYNC_USER_LIST_FILES', db.userFiles, room=sid)

    @staticmethod
    def updateEmojiData(sid=None):
        sio.emit('SYNC_EMOJI_DATA', db.emojiList, room=sid)

    @staticmethod
    def updateAdminStatus(admin, sid=None):
        sio.emit('SYNC_ADMIN', admin, room=sid)

    @staticmethod
    def updateGameSettings(settings, sid=None):
        sio.emit('SYNC_SETTINGS', settings, room=sid)

    @staticmethod
    def updatePlayerData(data, sid=None):
        sio.emit('SYNC_PLAYERS', data, room=sid)

    @staticmethod
    def updateSongList(sid=None):
        sio.emit('SYNC_FULL_LIST', db.songList, room=sid)

    @staticmethod
    def updateGameChoices(sid=None):
        sio.emit('SYNC_CHOICES', db.choices, room=sid)

    @staticmethod
    def updateGameStatePlaying(playing, sid=None):
        sio.emit('SYNC_PLAYING', playing, room=sid)

    @staticmethod
    def updateGameStateSongCount(count, sid=None):
        sio.emit('SYNC_SONG_COUNT', count, room=sid)

    @staticmethod
    def updateGameStateNewSong(song, position, sid=None):
        sio.emit('NEW_SONG', data=(song, position), room=sid)

    @staticmethod
    def updateGameStateStartCountdown(guessTime, sid=None):
        sio.emit('START_COUNTDOWN', guessTime, room=sid)

    @staticmethod
    def updateGameStateTimeUp(sid=None):
        sio.emit('TIME_UP', room=sid)

    @staticmethod
    def updateGameStateShowGuess(sid=None):
        sio.emit('SHOW_GUESS', room=sid)

    @staticmethod
    def updateSongSelector(sid=None):
        sio.emit('RESET_SELECTOR', room=sid)

File: server_ts/src/shared/handlers.ts
import * as socketio from 'socket.io'
import { AMQDatabase } from '../database/index'
import { db } from '../init'
import { PlayerDataSerialized, SettingObj, SongCount, Song } from './interfaces'

class MessageEmitter {
  _io: socketio.Server
  _db: AMQDatabase

  constructor(io: socketio.Server, db: AMQDatabase) {
    this._io = io
    this._db = db
  }

  notification(msgType: string, msg: string, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYSTEM_NOTIFICATION', msgType, msg)
  }

  syncSongList(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_FULL_LIST', this._db.songList)
  }

  syncUsers(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_USER_LIST_FILES', this._db.users)
  }

  syncUserList(user: string, socket: socketio.Socket=null): void {
    let data = {
      list: db.getUserList(user),
      file: user
    }
    this._emitter(socket).emit('SYNC_USER_LIST', data)
  }

  syncEmojiData(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_EMOJI_DATA', this._db.emojiList)
  }

  syncAdminStatus(admin: boolean, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_ADMIN', admin)
  }

  syncGameStatePlaying(playing: boolean, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_PLAYING', playing)
  }

  syncPlayerData(data: PlayerDataSerialized, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_PLAYERS', data)
  }

  syncGameChoices(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_CHOICES', this._db.choices)
  }

  syncGameSetting(settings: SettingObj, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_SETTINGS', settings)
  }

  syncSongCount(counts: SongCount, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SYNC_SONG_COUNT', counts)
  }

  gameNewSong(currentSong: Song, startPosition: number, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('NEW_SONG', currentSong, startPosition)
  }

  gameStartCountdown(guessTime: number, socket: socketio.Socket=null): void {
    this._emitter(socket).emit('START_COUNTDOWN', guessTime)
  }

  gameTimeUp(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('TIME_UP')
  }

  gameShowGuess(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('SHOW_GUESS')
  }

  gameReset(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('RESET')
  }

  resetSelector(socket: socketio.Socket=null): void {
    this._emitter(socket).emit('RESET_SELECTOR')
  }

  gameSelectSong(sid: string, selectTime: number): void {
    let client = this._getClient(sid)
    if (client) {
      client.emit('SELECT_SONG', selectTime)
    }
  }

  gameSelectSongOver(sid: string): void {
    let client = this._getClient(sid)
    if (client) {
      client.emit('SELECT_SONG_OVER')
    }
  }

  _getClient(sid: string): socketio.Socket {
    return this._io.nsps['/'].connected[sid]
  }

  _emitter(socket: socketio.Socket=null): socketio.Socket | socketio.Server {
    if (socket) {
      return socket
    }
    return this._io
  }
}

export { MessageEmitter }